lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -579,6 +587,41 @@ class TypeRefinementContextBuilder : private ASTWalker {
     return E;
   }
 };
+
+void TypeRefinementContextBuilder::buildCaseStmtRefinementContext(CaseStmt *CS) {
+  Optional<AvailabilityContext> BodyRange = None;
+  for (const auto &caseItem : CS->getCaseLabelItems()) {
+    auto *elementPattern = dyn_cast<EnumElementPattern>(caseItem.getPattern());
+    if (!elementPattern)
+      continue;
+
+    auto *elementDecl = elementPattern->getElementDecl();
+    Decl *D = elementDecl;
+    // If enum element has available attribute then infer context from it
+    // Otherwise inherit context from enum decl
+    if (!hasActiveAvailableAttribute(D, Context)) {
+      D = elementDecl->getParentEnum();
+    }
+
+    AvailabilityContext ElementInfo = swift::AvailabilityInference::availableRange(D, Context);
+    if (BodyRange.hasValue()) {
+      // If case stmt has multiple labels we unite their contexts
+      BodyRange.getValue().unionWith(ElementInfo);
```

Comment:
Don't we actually want the most restrictive refinement context - the intersection?  Suppose you have

```
switch foo {
case .onlyAvailableOnDishwasherOS1200, .onlyAvailableOnDishwasherOS1300:
 // Refines to DishwasherOS1300
```

Reply:
Please also include this in a test.

Reply:
You are right! Im my first implementation there was intersection and then I got a little lost. There are test case for multiple case labels. Will redo to intersection.

Reply:
Done

Reply:
No, I still think there must be the weakest refinement context. Here is counterexample for intersection:
```
func handleEnum(e: Enum) {
  switch e {
  case .onlyAvailableOnDishwasherOS1200, .onlyAvailableOnDishwasherOS1300:
  // Refines to DishwasherOS1300
  }
}
handleEnum(e: .onlyAvailableOnDishwasherOS1200)
```
If we run that on _DishwasherOS1200_ we get refined context to _DishwasherOS1300_ without any warnings

Reply:
Ack, of course I had the right words but the wrong idea - lattices are fun. Yes.  We have to treat multiple cases as if there is a false flow for each one and therefore we must take the most restrictive context: the *least* availability version. And how do we take the least?  With the union. 

Sorry for this.